### PR TITLE
Add backend API for implicit state

### PR DESCRIPTION
**Pull request by Tabrizian**
The backend should call the new APIs in the order below:

1. TRITONBACKEND_StateNew
2. TRITONBACKEND_StateBuffer
3. TRITONBACKEND_StateUpdate

### CODE REVIEW

File: include/triton/core/tritonbackend.h [hunk 1]
```diff
@@ -580,6 +603,39 @@ TRITONBACKEND_DECLSPEC TRITONSERVER_Error* TRITONBACKEND_ResponseOutput(
...
+///
+/// \param response The request.
+/// \param state Returns the new response output state.
+/// \param name The name of the output tensor.
+/// \param datatype The datatype of the output tensor.
+/// \param shape The shape of the output tensor.
+/// \param dims_count The number of dimensions in the output tensor
+/// shape.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONBACKEND_DECLSPEC TRITONSERVER_Error* TRITONBACKEND_RequestStateNew(
```
Comment (deadeyegoodwin) [inline include/triton/core/tritonbackend.h:618]: Think of each state as similar to a response. So have TRITONBACKEND_StateNew, TRITONBACKEND_StateDelete and TRITONBACKEND_StateUpdate. Be sure to document lifecycle and ownership of State objects.

Make a TRITONBACKEND_State section for all the state related functions

**Review by deadeyegoodwin — CHANGES_REQUESTED** [review]

File: include/triton/core/tritonbackend.h [hunk 2]
```diff
@@ -580,6 +603,39 @@ TRITONBACKEND_DECLSPEC TRITONSERVER_Error* TRITONBACKEND_ResponseOutput(
...
+    TRITONBACKEND_Request* request, TRITONBACKEND_State** state,
+    const char* name, const TRITONSERVER_DataType datatype,
+    const int64_t* shape, const uint32_t dims_count);
+
+/// Update the implicit state stored for each sequence. If the number of state
```
Comment (tanmayv25) [inline include/triton/core/tritonbackend.h:623]: So, the state here is a collection of states described in the protobuf? Looks bit strange..

File: include/triton/core/tritonbackend.h [hunk 3]
```diff
@@ -321,6 +322,28 @@ TRITONBACKEND_DECLSPEC TRITONSERVER_Error* TRITONBACKEND_OutputBuffer(
...
     int64_t* memory_type_id);
 
+/// Get a buffer to use to hold the tensor data for the state output. The
+/// returned buffer is owned by the output and so should not be freed by the
+/// caller. The caller can and should fill the buffer with the output state. The
+/// buffer will no longer be accesible to the backend after
+/// TRITONBACKEND_RequestStateUpdate is called. The caller should fill the
+/// buffer before calling TRITONBACKEND_RequestStateUpdate.
+///
+/// \param buffer Returns a pointer to a buffer where the contents of
```
Comment (tanmayv25) [inline include/triton/core/tritonbackend.h:332]: missing state description.

File: include/triton/core/tritonbackend.h (hunk 2 above)
Comment (Tabrizian) [reply include/triton/core/tritonbackend.h:623]: No, the state here is very similar to `TRITONBACKEND_ResponseOutput`. The protobuf description is only used for determing the shape and I/O mapping.

File: include/triton/core/tritonbackend.h [hunk 4]
```diff
@@ -595,6 +596,65 @@ TRITONBACKEND_DECLSPEC TRITONSERVER_Error* TRITONBACKEND_ResponseSend(
...
+/// of the model configuration, this call will not do anything.
+///
+/// \param state The state.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONBACKEND_DECLSPEC TRITONSERVER_Error* TRITONBACKEND_StateUpdate(
+    TRITONBACKEND_State* state);
+
+/// Get a buffer to use to hold the tensor data for the state. The returned
+/// buffer is owned by the state and so should not be freed by the caller. The
+/// caller can and should fill the buffer with the state buffer. The buffer will
```
Comment (deadeyegoodwin) [inline include/triton/core/tritonbackend.h:638]: "... with the state data".

File: include/triton/core/tritonbackend.h [hunk 5]
```diff
@@ -595,6 +596,65 @@ TRITONBACKEND_DECLSPEC TRITONSERVER_Error* TRITONBACKEND_ResponseSend(
...
+/// no longer be accesible to the backend after TRITONBACKEND_UpdateState
```
Comment (deadeyegoodwin) [inline include/triton/core/tritonbackend.h:639]: The buffer must not be accessed by the backend after TRITONBACKEND_UpdateState is called.

File: include/triton/core/tritonbackend.h [hunk 6]
```diff
@@ -595,6 +596,65 @@ TRITONBACKEND_DECLSPEC TRITONSERVER_Error* TRITONBACKEND_ResponseSend(
...
+/// is called. The caller should fill the buffer before calling
+/// TRITONBACKEND_UpdateState.
+///
+/// \param buffer Returns a pointer to a buffer where the contents of the state
```
Comment (deadeyegoodwin) [inline include/triton/core/tritonbackend.h:643]: Need param state

File: include/triton/core/tritonbackend.h [hunk 7]
```diff
@@ -595,6 +596,65 @@ TRITONBACKEND_DECLSPEC TRITONSERVER_Error* TRITONBACKEND_ResponseSend(
...
+/// \param dims_count The number of dimensions in the output tensor
+/// shape.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONBACKEND_DECLSPEC TRITONSERVER_Error* TRITONBACKEND_StateNew(
+    TRITONBACKEND_State** state, TRITONBACKEND_Request* request,
+    const char* name, const TRITONSERVER_DataType datatype,
+    const int64_t* shape, const uint32_t dims_count);
+
+/// Update the state for the sequence. Calling this function will replace the
+/// state stored for this seqeunce in Triton with the state provided in the
```
Comment (deadeyegoodwin) [inline include/triton/core/tritonbackend.h:626]: ...sequence in Triton with 'state'.

You should say something about how it is not required that the backend call this function and if the backend doesn't call this function then the state will not be updated and the next inference request in the sequence will use the existing state... or something like that.

Comment (Tabrizian) [reply include/triton/core/tritonbackend.h]: Added a sentence about this.

**Review by deadeyegoodwin — APPROVED** [review]

**Review by tanmayv25 — APPROVED** [review]

**Pull request merged**